core/src/main/java/org/apache/iceberg/BaseRemoveUnusedSpecs.java

```diff
@@ -0,0 +1,106 @@
+import static org.apache.iceberg.TableProperties.COMMIT_TOTAL_RETRY_TIME_MS;
+import static org.apache.iceberg.TableProperties.COMMIT_TOTAL_RETRY_TIME_MS_DEFAULT;
+
+import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors;
+import org.apache.iceberg.exceptions.CommitFailedException;
+import org.apache.iceberg.relocated.com.google.common.collect.Iterables;
+import org.apache.iceberg.relocated.com.google.common.collect.Sets;
+import org.apache.iceberg.util.Tasks;
+
+/**
+ * Implementation of RemoveUnusedSpecs API to remove unused partition specs.
+ *
+ * <p>When committing, these changes will be applied to the latest table metadata. Commit conflicts
+ * will be resolved by recalculating which specs are no longer in use again in the latest metadata
+ * and retrying.
+ */
+class BaseRemoveUnusedSpecs implements RemoveUnusedSpecs {
+  private final TableOperations ops;
+  private final Table table;
+
+  private TableMetadata base;
+
+  BaseRemoveUnusedSpecs(TableOperations ops, Table table) {
+    this.ops = ops;
+    this.base = ops.current();
+    this.table = table;
+  }
+
+  @Override
+  public List<PartitionSpec> apply() {
+    TableMetadata newMetadata = internalApply();
+    return newMetadata.specs();
+  }
+
+  @Override
+  public void commit() {
+    this.base = ops.refresh();
+    Tasks.foreach(ops)
+        .retry(base.propertyAsInt(COMMIT_NUM_RETRIES, COMMIT_NUM_RETRIES_DEFAULT))
+        .exponentialBackoff(
+            base.propertyAsInt(COMMIT_MIN_RETRY_WAIT_MS, COMMIT_MIN_RETRY_WAIT_MS_DEFAULT),
+            base.propertyAsInt(COMMIT_MAX_RETRY_WAIT_MS, COMMIT_MAX_RETRY_WAIT_MS_DEFAULT),
+            base.propertyAsInt(COMMIT_TOTAL_RETRY_TIME_MS, COMMIT_TOTAL_RETRY_TIME_MS_DEFAULT),
+            2.0 /* exponential */)
+        .onlyRetryOn(CommitFailedException.class)
+        .run(
+            taskOps -> {
+              TableMetadata newMetadata = internalApply();
+              taskOps.commit(base, newMetadata);
+            });
+  }
+
+  private static Iterable<ManifestFile> reachableManifests(Table table) {
+    Iterable<Snapshot> snapshots = table.snapshots();
+    Iterable<Iterable<ManifestFile>> manifestIterables =
+        Iterables.transform(snapshots, snapshot -> snapshot.allManifests(table.io()));
+    return Iterables.concat(manifestIterables);
+  }
+
+  private TableMetadata internalApply() {
+    this.base = ops.refresh();
+    List<PartitionSpec> specs = base.specs();
+    int currentSpecId = base.defaultSpecId();
+
+    Set<Integer> specsInUse =
+        Sets.newHashSet(
+            Iterables.transform(reachableManifests(table), ManifestFile::partitionSpecId));
+
+    specsInUse.add(currentSpecId);
+
+    List<PartitionSpec> specsToRemove =
+        specs.stream()
+            .filter(spec -> !specsInUse.contains(spec.specId()))
+            .collect(Collectors.toList());
+
+    return base.pruneUnusedSpecs(specsToRemove);
```

Comment:
This should use the `TableMetadata.Builder` rather than adding an operation to `TableMetadata` itself.